### PR TITLE
Issue#810 availabilities

### PR DESCRIPTION
**Pull request by Maxstgt**
Implemented @jenna1k and @VicoDraye components.
Implemented form which lets the buyer choose what part of the movie he wants to pay for.

### CODE REVIEW

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/+state/basket.service.ts [hunk 1]
```diff
@@ -1,16 +1,16 @@
...
   constructor(private db: FireQuery, private organizationQuery: OrganizationQuery) {}
 
-  public add(distributionRight: DistributionRight) {
-    const organizationId = this.organizationQuery.getValue();
+  public add(distributionRight: CatalogBasket) {
+    const organizationId = this.organizationQuery.getValue().org.id;
     this.db
-      .doc<DistributionRight>(`orgs/${organizationId}/distributionRight`)
-      .set(distributionRight);
+      .collection<CatalogBasket>(`orgs/${organizationId}/distributionRights`)
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/+state/basket.service.ts]: I was not quite sure, how to sort it. Valid point, but do we want to have a subcollection called basket? or Catalogbasket?

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.component.ts [hunk 2]
```diff
@@ -11,15 +26,249 @@ import { ChangeDetectionStrategy } from '@angular/core';
...
+  public subtitlesFilter: Observable<string[]>;
+  public selectedLanguages: string[] = [];
+  public selectedDubbings: string[] = [];
+  public selectedSubtitles: string[] = [];
+
+  // Datepicker section
+  public disabledDates: Date;
+  public matcher = new MyErrorStateMatcher();
+
+  // Territory of territory
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.component.ts]: I allowed me a small joke ;)

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.component.ts [hunk 3]
```diff
@@ -11,15 +26,249 @@ import { ChangeDetectionStrategy } from '@angular/core';
...
+          dubbed: data.dubbings[0],
+          subtitle: data.subtitles[0],
+          rights: data.medias[0]
+        }
+      ];
+      this.choosenDateRange.to = data.duration.to;
+      this.choosenDateRange.from = data.duration.from;
+    });
+    this.occupiedDateRanges.push({
+      to: (this.query.getActive().salesAgentDeal.rightsEnd as any).to.toDate(),
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.component.ts]: This data will be passed to Victors component, and I guess he only can handle `Date`. And I wouldn't change the Movie Interface now

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.form.ts [hunk 4]
```diff
@@ -18,25 +22,69 @@ export class DistributionRightForm extends FormEntity<DistributionRight> {
     });
   }
 
-  get medias() {
-    return this.get('medias');
-  }
-
-
-/*   addTerritory(territory: string) {
+  addTerritory(territory: string) {
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.form.ts]: I totally agree with you on that. But this whole FormEntity stuff from Francois there is no public and private. Also, I thought, maybe there is a deeper meaning here :D

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.html [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+              [indeterminate]="selection.hasValue() && !isAllSelected"
+              aria-label="select all movies">
+              </mat-checkbox>
+            </th>
+            <td mat-cell *matCellDef="let element">
+              <mat-checkbox
+              (click)="$event.stopPropagation()"
+              (change)="toggle(element)"
+              [checked]="selection.isSelected(element)"
+              aria-label="Select one Movie">
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.html]: This is a component from @jenna1k. I don't want to touch it. But if it's necessary I will change it.

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.ts [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+      // We listen on the change of selection and forward it to the parent
+      this.subscription = this.selection.changed.subscribe(() => this.selectionChange.emit(this.selection.selected));
+    }
+  
+    ngOnDestroy() {
+      // We stop listening when component is destroyed
+      this.subscription.unsubscribe();
+    }
+  
+    remove(movie: MovieData) {
```
[the remark this answers is not in the document]
Comment (Maxstgt) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.ts]: I start wondering who reviewed @jenna1k PR :D

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/+state/basket.service.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ioaNikas) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/+state/basket.service.ts]: I just looked the data model for catalog-app and it catalogBasket is just an array in organization.
So you can't set it as a sub-collection as an organization can only have 1 basket. You should just set the basket in organization :

`this.db.doc<Organization>(`orgs/${organizationId}`).update( { catalog:  distributionRight } )`

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.form.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (ioaNikas) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/create/create.form.ts]: Oh ok you maybe should ask him what is the correct way to do (as he told me to use prefixes on function long ago).

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.html (hunk 5 above)
[the remark this answers is not in the document]
Comment (ioaNikas) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.html]: I think you can change it, it's minor.

File: apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.ts (hunk 6 above)
[the remark this answers is not in the document]
Comment (ioaNikas) [reply apps/catalog-marketplace/catalog-marketplace/src/app/distribution-right/selection-table/form-selection.component.ts]: As I said above, keep it like this and ask @GrandSchtroumpf what is the correct way to declare functions before adding (or not) prefixes :p

**Pull request merged**